File: crack-data-structures-and-algorithms/leetcode/reverse_linked_list_q206.cpp
#include <utility>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 核心思路
// pp, p, pn 经典三指针
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        decltype(head) pp = nullptr;
        auto p = head;
        while (p) {
            auto pn = p->next;
            p->next = pp;
            pp = p;
            p = pn;
        }

        return pp;
    }
};

// 递归版本
// 核心思路
// reverse()函数返回新的head, tail指针对
// 保持这个定义就可以
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        auto cp = reverse(head);
        return cp.first;
    }

    // new head & tail respectively
    pair<ListNode*, ListNode*> reverse(ListNode* head) {
        if (!head || !head->next) {
            return {head, head};
        }

        auto cp = reverse(head->next);
        cp.second->next = head;
        head->next = nullptr;
        return {cp.first, head};
    }
};

File: crack-data-structures-and-algorithms/leetcode/python-impl/reverse_linked_list_q206.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # Special traits:
        # This question is whole-list reverse, thus
        # * No need for dummy node, since the final value of pp is our new head.
        # * No need to handle special cases like head == None or only a single node,
        #   because our loop invariant takes care of it well.
        pp = None
        p = head
        while p:
            pn = p.next
            p.next = pp
            pp = p
            p = pn

        return pp

# Use recursion

class SolutionRecursion(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        tail, root = reverse(head)
        return root

def reverse(head):
    if not head or not head.next:
        return head, head

    pp, root = reverse(head.next)
    pp.next = head
    head.next = None
    return head, root
